src/libcrun/linux.c

```diff
@@ -1812,22 +1813,93 @@ create_missing_devs (libcrun_container_t *container, bool binds, libcrun_error_t
   return 0;
 }
 
+static int
+get_default_flags (libcrun_container_t *container, const char *destination, char **data)
+{
+  if (strcmp (destination, "/proc") == 0)
+    return 0;
+  if (strcmp (destination, "/dev/cgroup") == 0 || strcmp (destination, "/sys/fs/cgroup") == 0)
+    {
+      *data = xstrdup ("none,name=");
+      return MS_NOEXEC | MS_NOSUID | MS_STRICTATIME;
+    }
+  if (strcmp (destination, "/dev") == 0)
+    {
+      *data = xstrdup ("mode=755");
+      return MS_NOEXEC | MS_STRICTATIME;
+    }
+  if (strcmp (destination, "/dev/shm") == 0)
+    {
+      *data = xstrdup ("mode=1777,size=65536k");
+      return MS_NOEXEC | MS_NOSUID | MS_NODEV;
+    }
+  if (strcmp (destination, "/dev/mqueue") == 0)
+    return MS_NOEXEC | MS_NOSUID | MS_NODEV;
+  if (strcmp (destination, "/dev/pts") == 0)
+    {
+      if (container->host_uid == 0)
+        *data = xstrdup ("newinstance,ptmxmode=0666,mode=620,gid=5");
+      else
+        *data = xstrdup ("newinstance,ptmxmode=0666,mode=620");
+      return MS_NOEXEC | MS_NOSUID;
+    }
+  if (strcmp (destination, "/sys") == 0)
+    return MS_NOEXEC | MS_NOSUID | MS_NODEV;
+
+  return 0;
+}
+
 static int
 do_masked_and_readonly_paths (libcrun_container_t *container, libcrun_error_t *err)
 {
-  size_t i;
+  size_t i, j, m;
+  char *data, *base_dir;
+  unsigned long flags, extra_flags;
+  uint64_t rec_clear, rec_set;
+  char readonly_path[PATH_MAX];
   int ret;
   runtime_spec_schema_config_schema *def = container->container_def;
 
   for (i = 0; i < def->linux->masked_paths_len; i++)
     {
-      ret = do_masked_or_readonly_path (container, def->linux->masked_paths[i], false, err);
+      ret = do_masked_or_readonly_path (container, def->linux->masked_paths[i], false, 0, err);
       if (UNLIKELY (ret < 0))
         return ret;
     }
   for (i = 0; i < def->linux->readonly_paths_len; i++)
     {
-      ret = do_masked_or_readonly_path (container, def->linux->readonly_paths[i], true, err);
+      data = NULL;
+      flags = 0;
+      extra_flags = 0;
+      rec_clear = 0;
+      rec_set = 0;
+      readonly_path[0] = '\0';
+
+      // Before setting path to readonly we must check if any exisitng mount is a parent
+      // if yes the we must inherit the flags/options from the parent.
+      strcpy (readonly_path, def->linux->readonly_paths[i]);
+      base_dir = dirname (readonly_path);
```

Comment:
I don't think we need to look up this information in the config spec but we need to use statfs to get info about the destination directory and see what flags are already set there.